sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,824 @@
+#endif
+  accessor<dataT,
+           dimensions,
+           accessmode,
+           accessTarget,
+           isPlaceholder,
+           property_list<Props...>> {/*...*/};
+
+} // namespace sycl
+```
+
+Illustrating this with the previous example:
+
+```
+namespace sycl {
+
+template </* ... */> class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_kernel_parameter_attributes(
+    "sycl-no-alias",  // Name of first property
+    "sycl-foo",       // Name of second property
+    nullptr,          // First property has no parameter
+    32                // Value of second property
+    )]]
+#endif
+  accessor</* ... */> {/* ... */};
+
+} // namespace sycl
+```
+
+As the name of the C++ attribute suggests, the device compiler front-end uses
+the attribute only when the decorated type is the type of a kernel argument,
+and it silently ignores the attribute when the class is used in any other way.
+
+When the device compiler front-end creates a kernel argument in this way, it
+adds one LLVM IR attribute to the kernel function's parameter for each property
+in the list.  For example, this can be done by calling
+[`Function::addParamAttrs(unsigned ArgNo, const AttrBuilder &)`][7].  As
+before, the IR attributes are added as strings, so the front-end must convert
+the property value to a string if it is not already a string.
+
+[7]: <https://llvm.org/doxygen/classllvm_1_1Function.html#a092beb46ecce99e6b39628ee92ccd95a>
+
+**TODO**: There are a number of open issues with this attribute and with the
+semantics of properties that are represented as attributes on kernel
+arguments.  Suppose there are two SYCL types that take properties: _A_ and
+_B_.  (For example, this could be two specializations of `annotated_ptr`, each
+decorated with different properties.)  Now suppose the application creates a
+struct that contains members with both of these types, and it passes that
+struct as a kernel argument.  What is the intended semantic?  Does the argument
+get decorated with the union of the properties on both _A_ and _B_?  What if
+those properties are mutually exclusive?  A similar case exists when the
+application creates a struct that inherits from both _A_ and _B_.
+
+The previous example shows a case when a single kernel argument gets properties
+from two (or more) types.  However, the opposite can also occur.  Certain SYCL
+classes are decorated with `__attribute__((sycl_special_class))`, which causes
+the compiler to pass each member of that class as a separate kernel argument.
+What should happen with the properties that decorate the class?  Should the
+compiler duplicate the properties on each such kernel argument?  Or, maybe it
+should be the header file's responsibility not to decorate such a class with
+`[[__sycl_detail__::add_ir_kernel_parameter_attributes()]]`, and instead it
+should decorate specific member variable(s) with this attribute?  How does the
+header decide which properties are used to decorate which member variables,
```

Comment:
While we don't want the frontend to have to have knowledge of what the properties mean, it's acceptable for the headers to have that knowledge.  After all the headers already need to encode which properties are valid on which classes in the is_property_of type traits.  And thus it's fine for the header to decide which properties map to which member variables on a case-by-case basis based on the meaning of the particular property.

Reply:
This is certainly an option.  If we go this route, we need to change the definition of `[[__sycl_detail__::add_ir_kernel_parameter_attributes()]]` so that it is allowed to decorate a member variable.  The header would then be responsible for decorating a single member variable of the class in cases where the class had `__attribute__((sycl_special_class))`.

Reply:
Getting back to this issue ... I propose we resolve this TODO by doing the following:

* Rename the C++ attribute to something like `[[__sycl_detail__::add_ir_special_kernel_parameter_attributes()]]` (note the addition of the word "special").
* Change the location of this attribute to decorate a member variable of a structure.
* Allow only the case when the attribute decorates a member variable of a "sycl special" structure.  (I.e. a struct or class that is decorated with `__attribute__((sycl_special_class))`.)

I think this covers the cases we really care about, and it solves both problems in the TODO.  In the case where the user  defines their own structure that encapsulates two "sycl special" types, the compiler will pass all members of both encapsulated types as separate kernel arguments.  Therefore, there is no longer a need to create a union of the properties because each member is decorated with at most one `[[__sycl_detail__::add_ir_special_kernel_parameter_attributes()]]` attribute.

The downside is that we can no longer support kernel argument properties on types that are **not** "sycl special".  I think this is OK, though.  Both the `accessor` and the `annotated_ptr` types are "sycl special", and I think those are the ones we care about in the short term.

@GarveyJoe (or others), any concerns?

Reply:
No concerns from me

Reply:
I updated the design to reflect my proposal above.  I did this in d0622a60026db7e2ff2d9a69ac8cdfb1ad05e917.  These TODO items are now removed.